store/postgres/src/entities.rs

```diff
@@ -191,105 +70,127 @@ pub(crate) fn make_storage_cache() -> StorageCache {
+            let layout = &self.data;
+            let start = Instant::now();
+            let base_layout =
+                &Connection::layout(&self.conn, base.namespace.clone(), &base.deployment)?;
+            layout.copy_from(
+                logger,
+                &self.conn,
+                &self.subgraph,
+                &base_layout,
+                &base.deployment,
+                block,
+            )?;
+            // Set the block ptr to the graft point to signal that we successfully
+            // performed the graft
+            deployment::forward_block_ptr(&self.conn, &self.subgraph, block.clone())?;
+            info!(logger, "Subgraph successfully initialized";
+            "time_ms" => start.elapsed().as_millis());
         }
-
-        // Clear the `migrating` lock on the subgraph; this flag must be
-        // visible to other db users before the migration starts and is
-        // therefore set in its own txn before migration actually starts.
-        // If the migration does not finish, e.g., because the server is shut
-        // down, `migrating` remains set to `true` even though no work for
-        // the migration is being done.
-        Ok(
-            diesel::update(dsl::table.filter(dsl::subgraph.eq(self.storage.subgraph.to_string())))
-                .set(dsl::migrating.eq(false))
-                .execute(self.conn.deref())
-                .map(|_| ())?,
-        )
+        Ok(())
     }
 
     pub(crate) fn find(
         &self,
-        entity: &String,
-        id: &String,
+        key: &EntityKey,
         block: BlockNumber,
     ) -> Result<Option<Entity>, StoreError> {
-        self.storage.find(&self.conn, entity, id, block)
+        assert_eq!(&self.subgraph, &key.subgraph_id);
+
+        match &key.entity_type {
+            EntityType::Data(name) => self.data.find(&self.conn, name, &key.entity_id, block),
+            EntityType::Metadata(typ) => {
+                METADATA_LAYOUT.find(&self.conn, typ.as_str(), &key.entity_id, block)
+            }
+        }
     }
 
     /// Returns a sequence of `(type, entity)`.
     /// If the entity isn't present that means it wasn't found.
     pub(crate) fn find_many(
         &self,
-        ids_for_type: BTreeMap<&str, Vec<&str>>,
+        ids_for_type: BTreeMap<&EntityType, Vec<&str>>,
         block: BlockNumber,
-    ) -> Result<BTreeMap<String, Vec<Entity>>, StoreError> {
-        self.storage.find_many(&self.conn, ids_for_type, block)
+    ) -> Result<BTreeMap<EntityType, Vec<Entity>>, StoreError> {
+        // Split the entities into data and metadata depending on their type
+        let data = ids_for_type
+            .iter()
+            .filter(|(typ, _)| typ.is_data_type())
+            .map(|(typ, ids)| (typ.as_str(), ids))
+            .collect();
+        let metadata = ids_for_type
+            .iter()
+            .filter(|(typ, _)| !typ.is_data_type())
+            .map(|(typ, ids)| (typ.as_str(), ids))
+            .collect();
```

Comment:
Looks like you want https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition.

Reply:
`partition` only lets you partition by a boolean predicate - I need to partition by the type name (i.e., potentially more than two partitions)

Reply:
I meant partitioning on `typ.is_data_type()`.

Reply:
It does seem tricky that you can't partition and map at the same time. It's fine to leave this as is.

Reply:
Aah, yes, I totally misunderstood; next time we need to touch this code ;)